tokio/src/util/slab.rs

```diff
@@ -0,0 +1,745 @@
+        // sees a slot pointer and initialized offset, it will remain valid
+        // until `compact()` is called. The `compact()` function also updates
+        // `CachedPage`.
+        unsafe {
+            let slot = self.slots.offset(idx as isize);
+            let value = slot as *const Value<T>;
+
+            &(*value).value
+        }
+    }
+}
+
+impl<T> Default for CachedPage<T> {
+    fn default() -> CachedPage<T> {
+        CachedPage {
+            slots: ptr::null(),
+            init: 0,
+        }
+    }
+}
+
+impl<T> Slots<T> {
+    /// Maps a slot pointer to an offset within the current page.
+    ///
+    /// # Panics
+    ///
+    /// panics if the provided slot pointer is not contained by the page.
+    fn index_for(&self, slot: *const Value<T>) -> usize {
+        use std::mem;
+
+        let base = &self.slots[0] as *const _ as usize;
+
+        assert!(base != 0, "page is unallocated");
+
+        let slot = slot as usize;
+        let width = mem::size_of::<Slot<T>>();
+
+        assert!(slot >= base, "unexpected pointer");
+
+        let idx = (slot - base) / width;
+        assert!(idx < self.slots.len() as usize);
+
+        idx
+    }
+}
+
+impl<T: Entry> Slot<T> {
+    // Generates a `Ref` for the slot. This involves bumping the page's ref count.
+    fn gen_ref(&self, page: &Arc<Page<T>>) -> Ref<T> {
+        // The ref holds a ref on the page.
+        mem::forget(page.clone());
+        let slot = self as *const Slot<T>;
+        let value = slot as *const Value<T>;
+
+        Ref { value }
+    }
+}
+
+impl<T> Value<T> {
+    // Release the slot, returning the `Arc<Page<T>>` logically owned by the ref.
+    fn release(&self) -> Arc<Page<T>> {
+        // Safety: called by `Ref`, which owns an `Arc<Page<T>>` instance.
+        let page = unsafe { Arc::from_raw(self.page) };
+        page.release(self as *const _);
+        page
+    }
+}
+
+impl Address {
+    fn page(self) -> usize {
+        // Since every page is twice as large as the previous page, and all page
+        // sizes are powers of two, we can determine the page index that
+        // contains a given address by shifting the address down by the smallest
+        // page size and looking at how many twos places necessary to represent
+        // that number, telling us what power of two page size it fits inside
+        // of. We can determine the number of twos places by counting the number
+        // of leading zeros (unused twos places) in the number's binary
+        // representation, and subtracting that count from the total number of
+        // bits in a word.
+        let slot_shifted = (self.0 + PAGE_INITIAL_SIZE) >> PAGE_INDEX_SHIFT;
+        (bit::pointer_width() - slot_shifted.leading_zeros()) as usize
+    }
+
+    pub(crate) const fn as_usize(self) -> usize {
+        self.0
+    }
+
+    pub(crate) fn from_usize(src: usize) -> Address {
+        Address(src)
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Might be nice to have some loom tests as well?

Reply:
Do you have thoughts on the specific behaviors you would want to see? The *majority* of the work happens via a mutex, there isn't too much to test. I can do some generic tests though.